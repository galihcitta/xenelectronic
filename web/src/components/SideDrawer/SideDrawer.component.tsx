import React from 'react';

import { List, Wrapper, Header, Logo, CloseBtn } from './SideDrawer.styles';

export interface SideDrawerProps {
  toggled: boolean;
  zIndex: number;
  title: string;
  handleClose: () => void;
  children: JSX.Element | JSX.Element[] | any;
}

export const SideDrawer: React.FC<SideDrawerProps> = ({
  toggled,
  zIndex,
  title,
  handleClose,
  children
}) => (
  <Wrapper data-testid="Wrapper" toggled={toggled} zIndex={zIndex}>
    <Header>
      <Logo>{title}</Logo>
      <CloseBtn title="close menu" aria-label="close" onClick={handleClose}>
        <i className="far fa-times-circle"></i>
      </CloseBtn>
    </Header>
    <List>{children}</List>
  </Wrapper>
);
