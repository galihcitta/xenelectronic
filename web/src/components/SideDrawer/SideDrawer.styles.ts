import styled from 'styled-components';

import { color, device, fontWeight } from '../../utils/styles';

interface IProps {
  zIndex: number;
  toggled: boolean;
}

export const Wrapper = styled.nav`
  position: fixed;
  z-index: ${(props: IProps) => props.zIndex};
  top: 0;
  left: 0;
  width: 100%;
  max-width: 260px;
  height: 100%;
  transform: ${(props: IProps) =>
    props.toggled ? 'translateX(0)' : 'translateX(-100%)'};
  transition: transform 0.2s ease;
  background: ${color.white};

  @media ${device.tablet} {
    max-width: 400px;
  }

  @media ${device.laptop} {
    display: none;
  }
`;

export const Header = styled.header`
  width: 100%;
  height: 75px;
  background: ${color.black};
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding-left: 0.5em;
`;

export const Logo = styled.p`
  font-size: 1rem;
  color: ${color.white};
  font-weight: ${fontWeight.bold};
  text-transform: uppercase;
`;

export const CloseBtn = styled.button`
  border: none;
  background: none;
  width: 50px;
  height: 50px;
  font-size: 1.55rem;
  color: ${color.white};
  cursor: pointer;
`;

export const List = styled.ul`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
`;
