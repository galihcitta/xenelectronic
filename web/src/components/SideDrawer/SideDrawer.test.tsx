import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { SideDrawer, SideDrawerProps } from './SideDrawer.component';

describe('<SideDrawer />', () => {
  const mockHandleClose = jest.fn();

  const props: SideDrawerProps = {
    toggled: false,
    zIndex: 1,
    title: 'Title',
    handleClose: mockHandleClose,
    children: <span>test</span>
  };

  const setup = (props: SideDrawerProps) => {
    const utils = render(<SideDrawer {...props} />);
    return { ...utils };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  it('contains correct style rules in `Wrapper` element', () => {
    const { getByTestId } = setup({ ...props, toggled: true, zIndex: 2 });
    const wrapper = getByTestId('Wrapper');

    expect(wrapper).toHaveStyle('z-index: 2');
    expect(wrapper).toHaveStyle('transform: translateX(0)');
  });

  it('calls `handleClose` prop', () => {
    const { getByLabelText } = setup(props);
    const btn = getByLabelText('close');

    fireEvent.click(btn);
    expect(mockHandleClose).toHaveBeenCalledTimes(1);
  });
});
