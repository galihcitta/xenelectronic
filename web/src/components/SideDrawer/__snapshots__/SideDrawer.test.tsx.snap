// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideDrawer /> renders correctly 1`] = `
<DocumentFragment>
  <nav
    class="sc-bdVaJa heZQzu"
    data-testid="Wrapper"
  >
    <header
      class="sc-bwzfXH gEGrQn"
    >
      <p
        class="sc-htpNat cuHHjG"
      >
        Title
      </p>
      <button
        aria-label="close"
        class="sc-bxivhb gkQASD"
        title="close menu"
      >
        <i
          class="far fa-times-circle"
        />
      </button>
    </header>
    <ul
      class="sc-ifAKCX hshtzb"
    >
      <span>
        test
      </span>
    </ul>
  </nav>
</DocumentFragment>
`;
